### PR TITLE
Compliance with catalog rules and sync quote items in cart estimate (#410)

### PR DESCRIPTION
**Pull request by mbouchotoyst**

### CODE REVIEW

File: app/code/community/Oyst/OneClick/Model/Catalog.php [hunk 1]
```diff
@@ -954,4 +961,54 @@ protected function getForbiddenSalesRulesActions()
...
+     * @param array $apiData
+     * @param Oyst_OneClick_Model_Magento_Quote $magentoQuoteBuilder
+     * @param OneClickOrderCartEstimate $oneClickOrderCartEstimate
+     */
+    protected function getCartItems($apiData, $magentoQuoteBuilder, $oneClickOrderCartEstimate)
+    {
+        $oystItems = array();
+
+        foreach ($magentoQuoteBuilder->getQuote()->getAllItems() as $item) {
+            if($item->getParentItemId()) {
```
Comment (mehdichaouch) [inline app/code/community/Oyst/OneClick/Model/Catalog.php]: run CSFIX add the end of your changes plz ;-)

File: app/code/community/Oyst/OneClick/Model/Catalog.php [hunk 2]
```diff
@@ -954,4 +961,54 @@ protected function getForbiddenSalesRulesActions()
...
     }
+
+     /**
+     * Get cart items.
+     *
+     * @param array $apiData
+     * @param Oyst_OneClick_Model_Magento_Quote $magentoQuoteBuilder
+     * @param OneClickOrderCartEstimate $oneClickOrderCartEstimate
+     */
+    protected function getCartItems($apiData, $magentoQuoteBuilder, $oneClickOrderCartEstimate)
```
Comment (mehdichaouch) [inline app/code/community/Oyst/OneClick/Model/Catalog.php]: It seams that ```$apiData``` is not used.
Why you don't use reference of ```$magentoQuoteBuilder``` and ```$oneClickOrderCartEstimate```?

File: app/code/community/Oyst/OneClick/Model/Catalog.php [hunk 3]
```diff
@@ -954,4 +961,54 @@ protected function getForbiddenSalesRulesActions()
...
+                $childItem = null;
+                foreach ($magentoQuoteBuilder->getQuote()->getAllItems() as $tmpItem) {
+                    if($tmpItem->getParentItemId() == $item->getId()) {
+                        $childItem = $tmpItem;
+                        break;
+                    }
+                }
+                $reference = $item->getProductId().';'.$item->getId().';'.$childItem->getProductId();
+            } else {
+                $reference = $item->getProductId().';'.$item->getId();
```
Comment (mehdichaouch) [inline app/code/community/Oyst/OneClick/Model/Catalog.php]: ```else``` is not mandatory.
You can move ```$reference = $item->getProductId().';'.$item->getId();``` this above ```if (in_array($item->getProductType(), array('configurable'))) {```

File: app/code/community/Oyst/OneClick/Model/Catalog.php [hunk 4]
```diff
@@ -954,4 +961,54 @@ protected function getForbiddenSalesRulesActions()
...
+     *
+     * @param array $apiData
+     * @param Oyst_OneClick_Model_Magento_Quote $magentoQuoteBuilder
+     * @param OneClickOrderCartEstimate $oneClickOrderCartEstimate
+     */
+    protected function getCartItems($apiData, $magentoQuoteBuilder, $oneClickOrderCartEstimate)
+    {
+        $oystItems = array();
+
+        foreach ($magentoQuoteBuilder->getQuote()->getAllItems() as $item) {
```
Comment (mehdichaouch) [inline app/code/community/Oyst/OneClick/Model/Catalog.php]: Rename ```$item``` as ```$quoteItem```

File: app/code/community/Oyst/OneClick/Model/Catalog.php [hunk 5]
```diff
@@ -954,4 +961,54 @@ protected function getForbiddenSalesRulesActions()
...
+            if($item->getParentItemId()) {
+                continue;
+            }
+
+            $oystPriceIncludingTaxes = $this->getOystPriceFromQuoteItem($item);
+
+            $reference = null;
+
+            if (in_array($item->getProductType(), array('configurable'))) {
```
Comment (mehdichaouch) [inline app/code/community/Oyst/OneClick/Model/Catalog.php]: change for:
```$item->getProduct()->isConfigurable())```

File: app/code/community/Oyst/OneClick/Model/Catalog.php [hunk 6]
```diff
@@ -954,4 +961,54 @@ protected function getForbiddenSalesRulesActions()
...
+                        break;
+                    }
+                }
+                $reference = $item->getProductId().';'.$item->getId().';'.$childItem->getProductId();
+            } else {
+                $reference = $item->getProductId().';'.$item->getId();
+            }
+
+            $oystItem = new OneClickItem(
+                $reference, $oystPriceIncludingTaxes, $item->getQty()
```
Comment (mehdichaouch) [inline app/code/community/Oyst/OneClick/Model/Catalog.php]: Each params per line or all in one line.

File: app/code/community/Oyst/OneClick/Model/Magento/Quote.php [hunk 7]
```diff
@@ -183,6 +183,10 @@ private function syncCustomer()
             $this->quote->setCheckoutMethod($checkoutMethod);
         }
 
+        if ($customer instanceof Mage_Customer_Model_Customer) {
+            Mae::getSingleton('customer/session')->setCustomer($customer);
```
Comment (mehdichaouch) [inline app/code/community/Oyst/OneClick/Model/Magento/Quote.php]: ```Mae```??? -_-'

File: app/locale/en_US/Oyst_OneClick.csv [hunk 8]
```diff
@@ -188,3 +188,5 @@
 
 "Allowed IPs (comma separated)","Allowed IPs (comma separated)"
 "Leave empty for access from any location.","Leave empty for access from any location."
+
+"Oyst has recognized you as a customer, so you benefit a promotion on this product.","Oyst has recognized you as a customer, so you benefit a promotion on this product."
```
Comment (mehdichaouch) [inline app/locale/en_US/Oyst_OneClick.csv]: Add new line add the end of file.

File: app/locale/fr_FR/Oyst_OneClick.csv [hunk 9]
```diff
@@ -188,3 +188,5 @@
 
 "Allowed IPs (comma separated)","Adresses IP autorisées (séparées par des virgules)"
 "Leave empty for access from any location.","Laissez vide pour l'accès de n'importe quel endroit."
+
+"Oyst has recognized you as a customer, so you benefit a promotion on this product.","Oyst vous a identifié en tant que client, vous bénéficiez d'une promotion sur ce produit."
```
Comment (mehdichaouch) [inline app/locale/fr_FR/Oyst_OneClick.csv]: Add new line add the end of file.

**Review by mehdichaouch — CHANGES_REQUESTED** [review]

**Review by mehdichaouch — APPROVED** [review]

**Review by StefBrito — APPROVED** [review]
Good pour moi

**Pull request merged**